### DISCUSSION TITLE
Trouble handling JS Object in Automation

### DISCUSSION THREAD

**Discussion by sandbak** [Budibase/budibase#6001] [Help] [answered]
BudiBase noob here, and by all means no more than a JS amateur.

I've got an automation set up that takes an object from an external datasource, which I then try to process with Javascript.

1. At the bottom here is the output of the external data connector step, which will be input to the next step.

2. I then add a 'create row' step, where one of the datafields should be filled with some value coming from the object in the output in step  one.

I can make this work locally on my PC with this data as an object, and a line of JS that selects the right value. I also succeeded in Budibase in returning a value that gets filled back into the row.

What I can't get to work though is how to pick up the object from step 1, and process that. I have tried in all kinds of ways to simply  pull some data form the object, and return it to BudiBase, but all I get is 'Error while executing JS'.

Most of what I tried is variations of this:

```
let obj = $("steps.1.response");                    // behind the = is the standard code that budibase offers to get the output as an object
let IntentLabels = obj.info['0'].rows[0].IntentLabel;      // to get the nested object with Intent Labels
let SampleReturn = Object.keys(IntentLabels)[0];     // just grabbing the first key as a sample string to feed back to test
return SampleReturn;
```

The first line by itself will execute without error. But the second line already causes a JS error. With that I am assuming the obj is not formatted or handled in the right way.

No matter how I try to simplify steps to just get anything from the object to process, I keep getting  'Error while executing JS'.

Given the lack of (as far as I know), options debug JS in budibase, I can't trace the steps.

Any thoughts on what I might be doing wrong??

```
{
  "info": {
    "0": {
      "rows": [
        {
          "IntentLabel": {
            "ContactCS": 0.00011947443,
            "UpdateEmailAddress": 0.00016584409,
            "AskRefundPolicy": 0.000018826886,
            "RetrieveInvoice": 0.00016305009,
            "AddressUpdate": 0.000049881695,
            "PasswordReset": 0.000038668153,
            "AskRegistration": 0.00017738089,
            "EditAccount": 0.000050853203,
            "FileComplaint": 0.00010118374,
            "DeleteAccount": 0.000051472027,
            "GetRefund": 0.00003842139,
            "PaymentIssue": 0.000119750264,
            "CreateAccount": 0.0001521948,
            "UnsubscribeMe": 0.99595946,
            "MakeOrder": 0.000325089,
            "TrackRefund": 0.000044837838,
            "AskSwitchAccount": 0.00003194304,
            "LeaveReview": 0.000062546984,
            "TrackOrder": 0.0003447139,
            "CouponIssue": 0.00026593058,
            "SubscribeMe": 0.001163405,
            "ChangeOrder": 0.0001359491,
            "DeliveryIssue": 0.00013375374,
            "AskDeliveryOptions": 0.00004644739,
            "AskPaymentMethods": 0.000018943936,
            "ContactHuman": 0.00021988149
          }
        }
      ]
    }
  },
  "success": true
}
```

**Comment by melohagan** [comment]
HI @sandbak 

This is the problem here:
```
"info": {
"0": {
"rows": [
```

Info is not an array. You'll need to do: `let IntentLabels = obj.info['0'].rows[0].IntentLabel; `

**Note the single quotes around the '0' in info**

**Reply by sandbak** [reply]
Hi @melohagan,

Thanks so much for having a look! I was actually aware that info wasn't an array, but got the code to work without the single quotes locallly on my PC. Wrong; but somehow working.. Thanks for pointing this out!

I've corrected it now following your example in Budibase, but unfortunately it still gives the error!

Any other thoughts?

(I'm correcting this in my original question so other don't replicate)

**Reply by melohagan** [reply]
Hmm maybe Budibase is fussy about the quotes. Do double quotes work? e.g. `...info["0"]...`

**Reply by sandbak** [reply]
Just tried, unfortunately that also doesn't work..

**Reply by sandbak** [reply]
to avoid the object / array confusion, I've tried the example below, which also throws an error:

```
let obj = $("steps.1.response");                   
let success = obj.success;   // get boolean from success key
let SuccessString = success.toString();  // convert boolean to string 
return SuccessString;
```

again the error already happens if I only add the second line, so there already something isn't working right..

**Comment by melohagan** [comment, marked as answer]
It appears that you need to use the info binding: `let IntentLabels = $("steps.1.info")[0].rows[0].IntentLabel;` 

This seems a bit strange given the binding descriptions, so I've raised a bug here: https://github.com/Budibase/budibase/issues/6003

**Reply by sandbak** [reply]
Awesome, that fixes things, thanks so much, @melohagan!

**Comment by Gogowitsch** [comment]
Is there a way to get the actual JS error instead of the generic error? A server log maybe?
